Configuration caches

ulmus-scott opened the pull request:
##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] contribution does not duplicate one of our [existing pull requests](https://github.com/MythTV/mythtv/pulls)
- [x] contribution is in a branch rebased against [master](https://github.com/MythTV/mythtv)
- [x] code compiles successfully without errors
- [x] code follows the [MythTV Coding Standards](https://www.mythtv.org/wiki/Coding_Standards)
- [x] commits are logically organised and have [good commit messages](https://chris.beams.io/posts/git-commit)



@stuarta I had hoped to do more work on this, regarding the XmlConfigurations use, but this should be an improvement over what you did.

The configuration caches are currently in libmyth/mythcontext (private).

See the commit messages for details.  None of the UPnP stuff was ever saved.

ulmus-scott commented:
Compiles, but not working.  I'm investigating.

stuarta commented on mythtv/libs/libmythupnp/eventing.cpp:
```diff
@@ -75,7 +77,7 @@ Eventing::Eventing(const QString &sExtensionName,
     m_sEventMethodName(std::move(sEventMethodName)),
     m_nSubscriptionDuration(
-        UPnp::GetConfiguration()->GetDuration<std::chrono::seconds>("UPnP/SubscriptionDuration", 30min))
+        GetMythDB()->GetDuration<std::chrono::seconds>("UPnP/SubscriptionDuration", 30min))
```
This isn't correct. The database does not contain any setting for "UPnP/SubscriptionDuration". This is *only* (optionally) present in config.xml.

ulmus-scott replied:
The [wiki](https://www.mythtv.org/wiki/Config.xml) doesn't mention these settings, so I thought they would be in the database.  Only UPnP/UDN/MediaRenderer was ever saved by the code.

Should these settings be saved to config.xml?

ulmus-scott replied:
Since this is read only, would `XmlConfiguration("config.xml").Get …` or equivalent be acceptable?
```
auto config = XmlConfiguration("config.xml");
foobar = config.Get …
```

stuarta commented on mythtv/libs/libmythupnp/httprequest.cpp:
```diff
@@ -1860,7 +1861,7 @@ QString HTTPRequest::GetETagHash(const QByteArray &data)
 bool HTTPRequest::IsUrlProtected( const QString &sBaseUrl )
 {
-    QString sProtected = UPnp::GetConfiguration()->GetValue( "HTTP/Protected/Urls", "/setup;/Config" );
+    QString sProtected = GetMythDB()->GetSetting("HTTP/Protected/Urls", "/setup;/Config");
```
Again, not a DB setting

stuarta commented on mythtv/libs/libmythupnp/mmulticastsocketdevice.cpp:
```diff
@@ -44,7 +45,7 @@ MMulticastSocketDevice::MMulticastSocketDevice(
 {
 #if 0
-    ttl = UPnp::GetConfiguration()->GetValue( "UPnP/TTL", 4 );
+    ttl = GetMythDB()->GetNumSetting("UPnP/TTL", 4); // also uncomment mythdb.h above
```
Again not a DB setting

stuarta commented on mythtv/libs/libmythupnp/ssdp.cpp:
```diff
@@ -75,10 +78,8 @@ SSDP::SSDP() :
-    m_nPort       = pConfig->GetValue("UPnP/SSDP/Port"      , SSDP_PORT      );
-    m_nSearchPort = pConfig->GetValue("UPnP/SSDP/SearchPort", SSDP_SEARCHPORT);
+    m_nPort       = GetMythDB()->GetNumSetting("UPnP/SSDP/Port"      , SSDP_PORT      );
+    m_nSearchPort = GetMythDB()->GetNumSetting("UPnP/SSDP/SearchPort", SSDP_SEARCHPORT);
```
Not a DB setting

stuarta commented on mythtv/libs/libmythupnp/ssdp.cpp:
```diff
@@ -691,8 +692,7 @@ SSDPExtension::SSDPExtension( int nServicePort , const QString &sSharePath)
     m_nSupportedMethods |= (RequestTypeMSearch | RequestTypeNotify);
-    m_sUPnpDescPath = UPnp::GetConfiguration()->GetValue( "UPnP/DescXmlPath",
-                                                 m_sSharePath );
+    m_sUPnpDescPath = GetMythDB()->GetSetting("UPnP/DescXmlPath", m_sSharePath);
```
Not a DB setting

stuarta commented on mythtv/libs/libmythupnp/upnpcds.cpp:
```diff
@@ -92,7 +93,7 @@ UPnpCDS::UPnpCDS( UPnpDevice *pDevice, const QString &sSharePath )
                           QDateTime::currentDateTimeUtc().toString(Qt::ISODate) );
 
-    QString sUPnpDescPath = UPnp::GetConfiguration()->GetValue( "UPnP/DescXmlPath", sSharePath );
+    QString sUPnpDescPath = GetMythDB()->GetSetting("UPnP/DescXmlPath", sSharePath);
```
Not a DB setting

stuarta commented on mythtv/libs/libmythupnp/upnpcmgr.cpp:
```diff
@@ -34,8 +35,7 @@ UPnpCMGR::UPnpCMGR ( UPnpDevice *pDevice,
 
-    QString sUPnpDescPath = UPnp::GetConfiguration()->GetValue( "UPnP/DescXmlPath",
-                                                                m_sSharePath );
+    QString sUPnpDescPath = GetMythDB()->GetSetting("UPnP/DescXmlPath", m_sSharePath );
```
Not a DB setting

stuarta commented on mythtv/libs/libmythupnp/upnpdevice.cpp:
```diff
@@ -361,8 +361,7 @@ void UPnpDeviceDesc::OutputDevice( QTextStream &os,
     if (pDevice == &m_rootDevice)
-        sFriendlyName = UPnp::GetConfiguration()->GetValue( "UPnP/FriendlyName",
-                                                            sFriendlyName  );
+        sFriendlyName = GetMythDB()->GetSetting("UPnP/FriendlyName", sFriendlyName);
```
Not a DB setting

stuarta commented on mythtv/libs/libmythupnp/upnpmsrr.cpp:
```diff
@@ -33,8 +34,7 @@ UPnpMSRR::UPnpMSRR( UPnpDevice *pDevice, const QString &sSharePath )
 
-    QString sUPnpDescPath =
-        UPnp::GetConfiguration()->GetValue( "UPnP/DescXmlPath", m_sSharePath );
+    QString sUPnpDescPath = GetMythDB()->GetSetting("UPnP/DescXmlPath", m_sSharePath);
```
Not a DB setting

stuarta commented on mythtv/libs/libmythupnp/upnptaskcache.h:
```diff
@@ -43,7 +45,7 @@ class SSDPCacheTask : public Task
             m_nExecuteCount = 0;
             m_nInterval     =
-                UPnp::GetConfiguration()->GetDuration<std::chrono::seconds>("UPnP/SSDP/CacheInterval", 30s);
+                GetMythDB()->GetDuration<std::chrono::seconds>("UPnP/SSDP/CacheInterval", 30s);
```
Not a DB setting

stuarta commented on mythtv/libs/libmythupnp/upnptasknotify.cpp:
```diff
@@ -44,8 +46,7 @@ UPnpNotifyTask::UPnpNotifyTask( int nServicePort ) :
     m_nServicePort = nServicePort;
-
-    m_nMaxAge      = UPnp::GetConfiguration()->GetDuration<std::chrono::seconds>( "UPnP/SSDP/MaxAge" , 1h );
+    m_nMaxAge = GetMythDB()->GetDuration<std::chrono::seconds>("UPnP/SSDP/MaxAge" , 1h);
```
Not a DB setting

stuarta commented on mythtv/libs/libmythupnp/upnptasksearch.cpp:
```diff
@@ -55,7 +56,7 @@ UPnpSearchTask::UPnpSearchTask( int          nServicePort,
     m_sUDN        = std::move(sUDN);
     m_nServicePort= nServicePort;
-    m_nMaxAge     = UPnp::GetConfiguration()->GetDuration<std::chrono::seconds>( "UPnP/SSDP/MaxAge" , 1h );
+    m_nMaxAge     = GetMythDB()->GetDuration<std::chrono::seconds>( "UPnP/SSDP/MaxAge" , 1h );
```
Not a DB setting

stuarta commented on mythtv/libs/libmythupnp/upnputil.cpp:
```diff
@@ -60,7 +62,7 @@ QString LookupUDN( const QString &sDeviceType )
     sList.removeLast();
     QString sName = "UPnP/UDN/" + sList.last();
-    QString sUDN  = UPnp::GetConfiguration()->GetValue( sName, "" );
+    QString sUDN  = GetMythDB()->GetSetting(sName, "");
```
Not a DB setting

ulmus-scott replied:
This was the only setting (UPnP/UDN/MediaRenderer) that was potentially saved.  Should it be saved?

It doesn't feel necessary to me to save this value, and doing so wreaks the read only nature of these UPnP settings.

If it should be saved, then should all of the UPnP settings should be in a separate upnp.xml file to preserve read only access to config.xml by the UPnP code?  (So as to not interfere with other code accessing config.xml.)

stuarta replied:
This is one of the few values that are actually of use. The majority of them probably don't even need to be configurable, and can be hard coded. We are currently undertaking surgical separation of the old webserver. the upnp and ssdp stacks and the config.xml handling code (this last bit I did prior to this PR), and this should be looked at as part of that.

ulmus-scott replied:
Does this need to be saved?

It is probably safe to do so, since XmlConfiguration reads in the entire file when created and this is the only code that looks at that value.

What do you think about my [proposed solution](https://github.com/MythTV/mythtv/pull/435#discussion_r775177963) for these not DB settings?

ulmus-scott replied:
> It is probably safe to do so, since XmlConfiguration reads in the entire file when created and this is the only code that looks at that value.

On a second examination, the setting could be lost in the following scenario:

1. MythContext loads config.xml first.
2. UPnP code loads config.xml.
3. UPnP code sets a UDN and saves it.
4. MythContext has changed database parameters, so it saves its new parameters, which erases the new UDN.

This was also present in the original code.  It is not present in your change; however, you introduced global state without locking.

stuarta commented on mythtv/programs/mythfrontend/mediarenderer.cpp:
```diff
@@ -40,17 +41,11 @@ MediaRenderer::MediaRenderer()
     // ----------------------------------------------------------------------
 
-    int nPort = g_pConfig->GetValue( "UPnP/MythFrontend/ServicePort", 6547 );
+    int nPort = GetMythDB()->GetNumSetting("UPnP/MythFrontend/ServicePort", 6547);
```
Not a DB setting

stuarta commented on mythtv/programs/mythfrontend/mythfexml.cpp:
```diff
@@ -35,8 +36,7 @@ MythFEXML::MythFEXML( UPnpDevice *pDevice , const QString &sSharePath)
 
-    QString sUPnpDescPath =
-        UPnp::GetConfiguration()->GetValue( "UPnP/DescXmlPath", m_sSharePath );
+    QString sUPnpDescPath = GetMythDB()->GetSetting("UPnP/DescXmlPath", m_sSharePath);
```
Not a DB setting

ulmus-scott replied:
Why wouldn't this be a DB setting like [mythtv/programs/mythbackend/mediaserver.cpp](https://github.com/MythTV/mythtv/pull/435/commits/2884f2d395aebf0b6192d70359fd7c3828df875b#diff-73a85b9f05d8f11a6e3c8cbcfc455d61149c46b6b1c8cf1bf2daec408149e597R99)?

stuarta replied:
The settings that are of relevance should be migrated to DB configuration items, unless they are required to gain access to get the config items out of the DB. This involves a read config.xml, and save to DB cycle rather than just changing the code to request everything out of the DB.

ulmus-scott commented on mythtv/libs/libmythupnp/configuration.cpp:
```diff
@@ -18,104 +8,85 @@
             LOG(VB_GENERAL, LOG_ERR,
                 QString("Error parsing: %1 at line: %2  column: %3")
-                    .arg( sName ) .arg( nErrLine ) .arg( nErrCol  ));
+                    .arg(pathName, line, column));
```
add explicit QString::number() for line and column for clarity

ulmus-scott commented on mythtv/libs/libmyth/mythcontext.cpp:
```diff
@@ -1242,8 +1248,8 @@ int MythContextPrivate::UPnPautoconf(const std::chrono::milliseconds milliSecond
-    QString pin = m_pConfig->GetValue(kDefaultPIN, "");
-    QString usn = m_pConfig->GetValue(kDefaultUSN, "");
+    QString pin = m_pConfig->GetValue(kDefaultPIN, QString("")); // not saved
+    QString usn = m_pConfig->GetValue(kDefaultUSN, QString("")); // not saved
```
These are potentially saved by https://github.com/MythTV/mythtv/blob/befa034b04d8449e0b1d41b0a28905cd9847dc37/mythtv/libs/libmyth/backendselect.cpp#L109-L119

TODO: rewrite this to eliminate pointer-y global state.  Because of how XmlCongifuration::Save() works, don't keep an object alive.

Instead:
1.  create a new object (on the stack)
2. Read whatever.
3. If writing, set (or clear) values, immediately save, and destroy the object.

ulmus-scott commented:
Rebased and force pushed.  (I meant to merge some of the commits together before pushing, but I forgot.  It is probably clearer while in progress to _not_ do that kind of rebasing, anyway.)

See https://github.com/MythTV/mythtv/pull/435/commits/8730698796c1c3ca1b0bae4fb1b1c7815da36e6c for my solution for the non-DB settings.  It works fine when I tested it.

Re: "UPnP/DescXmlPath" in mythfexml.cpp
stuarta:
>The settings that are of relevance should be migrated to DB configuration items, unless they are required to gain access to get the config items out of the DB. This involves a read config.xml, and save to DB cycle rather than just changing the code to request everything out of the DB.

I think that is beyond the scope of this pull request.

ulmus-scott commented:
Rebased onto master and added more logging.

ulmus-scott commented:
[WriteDelayedSettings.patch.txt](https://github.com/MythTV/mythtv/files/7857755/WriteDelayedSettings.patch.txt)  LOG_ALERT to make searching the log easier.
@stuarta
I could not verify that WriteDelayedSettings() works because I could not trigger appending a delayed setting.  This is probably because I was testing on a combined backend/frontend.

GetSettings() does return the default given.

WriteDelayedSettings() is called a number of times.  However, https://github.com/MythTV/mythtv/blob/4d3f06eec4fe8e9f3766d343c6c1c4b48acf18a3/mythtv/libs/libmythbase/mythdb.cpp#L966 should probably be renamed to IsCoreContextThread.

ulmus-scott commented:
Compilation failure from https://github.com/MythTV/mythtv/blob/377e00fa38fed9636e943eeeae80dda56aef5dd1/mythtv/programs/mythbackend/servicesv2/v2config.cpp#L80

Now converted to GetMythDB().

@stuarta The changes are otherwise identical.

Several commits will need to be squashed before merging this PR.

ulmus-scott commented:
Rebased onto master with the commits re-ordered and squashed.

The only difference is an extra newline in configuration.h and the removal of the void parameters in XmlConfiguration::(Save|Load).

ulmus-scott commented:
@stuarta I have now rebased this onto master and made a number of changes versus the previous version, including no longer using reversion commits.  The overall changes are similar.

ulmus-scott commented:
`configuration.h` was a public header when it was part of libmythupnp, and unless you see a reason it needs to be public, I'm fine keeping it a private header as it is currently in libmythbase.

ulmus-scott commented:
@stuarta Ping.  Rebased onto master to resolve minor conflict from minimum Qt version bump.

bennettpeter commented:
From the comments this pull seems to follow on to some discussion you had with stuart, which I know nothing about, so I need some context.

Please can you give an overall description of what the purpose is of the pull request, what features of MythTV are affected, and what changes will be visible to the user.

ulmus-scott commented:
> From the comments this pull seems to follow on to some discussion you had with stuart, which I know nothing about, so I need some context.

https://github.com/MythTV/mythtv/pull/435#issuecomment-1011455148 is about MythDB::WriteDelayedSettings(), which if I remember correctly, we discussed (privately?) on IRC.  Since 2010 with c93be1c6e497d97c9a568d06d56dc267e613b90f and 75cf4e4b4397576180f33f5c0b1e39717e76b083, you are supposed to be able to write settings to the database before it is connected.  Looking over the current changes, I don't think this is relevant to this pull request.

> Please can you give an overall description of what the purpose is of the pull request, 

I had started work on this before Stuart committed 6ba7c84c7e15d0c00741cdfa9fbcb8b8c5c07470 and the basic goal was the same: break dependencies on libmythupnp.

However, I went further than he did and also started some code cleanup.

My intention was to prepare to break the dependency of MythContext on libmythupnp, but I'm not sure that is completely possible.

> what features of MythTV are affected, 

Loading the configuration caches in `config.xml` for database settings and the frontend settings from `cache/contextcache.xml`.

> and what changes will be visible to the user.

Aside from the extra log messages I added, there should be no visible behavior changes.  UDNs for MediaRenderer, MediaServer, and MasterMediaServer will all end up saved in `config.xml`.  Only one of those is currently saved, I think.  I have those in my database's settings table because of an earlier version where I (incorrectly) made everything directly access the database.

bennettpeter reviewed (commented):
The destroying of the main window was to address the case where the xml cache had been deleted and you were starting up before the database was active, and got the wrong GUI setup in your window.

That destroying of the main window is no longer necessary, so this commit is good.

bennettpeter commented:
Merged into master.
